File: apps/web/src/utils/paymentMethod.js
// VIPOS — payment method canonical wire codes + display labels.
//
// Pairs with backend PR #232 (`apps/backend/src/lib/payment-methods.js`)
// and Android `PaymentMethod.kt`. The web kasir submits the canonical
// uppercase code (e.g. `CASH`, `EDC`, `QRIS_STATIC`) so the entire
// stack — Android, web kasir, backend, reports — speaks the same
// vocabulary. Legacy lowercase codes (`cash`, `card`, `qris`) are
// preserved in the display map so transactions saved before this
// migration still render with friendly Indonesian labels in the
// transactions list / receipts.

// UI id -> canonical wire code. The kasir UI state uses lowercase ids
// (so the existing cash-only branches `paymentMethod === 'cash'` keep
// working), but we translate to canonical at submit time. Adding a new
// kasir button just means adding a row here + a row in the methods
// array on `CashierPage.jsx`.
export const WIRE_CODE_FROM_UI_ID = Object.freeze({
  cash: 'CASH',
  card: 'EDC',
  qris: 'QRIS_STATIC',
});

// Display map for both legacy lowercase (rows saved before the
// migration) and canonical Android codes. Matches the Indonesian
// labels used by the kasir buttons + Android `PaymentMethod.kt`'s
// `displayName` resource so the UX is consistent.
const PAYMENT_METHOD_LABELS = Object.freeze({
  // Legacy lowercase (pre-#232 transactions).
  cash: 'Tunai',
  card: 'Kartu',
  qris: 'QRIS',
  // Canonical Android codes.
  CASH: 'Tunai',
  EDC: 'EDC',
  QRIS_STATIC: 'QRIS',
  QRIS_DYNAMIC: 'QRIS Dinamis',
  GOPAY: 'GoPay',
  OVO: 'OVO',
  DANA: 'DANA',
  SHOPEEPAY: 'ShopeePay',
  LINKAJA: 'LinkAja',
  BANK_TRANSFER: 'Transfer Bank',
  CREDIT: 'Kredit',
  DEPOSIT: 'Deposit',
  VOUCHER: 'Voucher',
  LOYALTY_POINT: 'Poin Loyalti',
  OTHER: 'Lainnya',
});

// Translate a kasir UI id to the canonical wire code. Falls back to
// the input verbatim so a future code passed through unchanged still
// reaches the backend (where the allow-list will reject anything
// unknown — see `lib/payment-methods.js`).
export function toWireCode(uiId) {
  if (typeof uiId !== 'string') return uiId;
  return WIRE_CODE_FROM_UI_ID[uiId] ?? uiId;
}

// Format any payment method code (legacy or canonical) for display.
// Falls back to the raw code if unknown so the UI never shows blank
// or `undefined` — the kasir always knows what was used to settle a
// transaction even if a future enum entry hasn't been added here.
export function formatPaymentMethodLabel(code) {
  if (typeof code !== 'string' || code.length === 0) return '—';
  return PAYMENT_METHOD_LABELS[code] ?? code;
}

File: apps/android/feature/pos/src/main/java/id/alviarts/vipos/feature/pos/domain/PaymentMethod.kt
package id.alviarts.vipos.feature.pos.domain

/**
 * Canonical inventory of payment methods the kasir can pick at
 * checkout (P3-08 first slice — data layer).
 *
 * The enum entries mirror the `code` column from
 * `docs/v2/14_PAYMENT_METHODS.md` §1 verbatim. The same code is
 * stored on the backend's `transactions.payment_method` column
 * and on every `transaction_payments` row, so any new entry here
 * MUST also be added to the backend's allow-list before merge —
 * otherwise the kasir can pick a method that the server rejects
 * at commit time. The current backend route
 * (`apps/backend/src/routes/transactions.js`) accepts an
 * arbitrary string and defaults to `cash`, so the enum is the
 * stricter contract until the backend gains its own validation.
 *
 * Methods that require a live network connection
 * ([requiresOnline] = `true`) are filtered out of the catalogue
 * when the device is offline; see [PaymentMethodCatalog]. Methods
 * that don't (cash, EDC manual, QRIS Statis, deposit, voucher,
 * loyalty point, custom) stay available even on a flaky link so
 * the kasir can keep ringing transactions through.
 *
 * Per-merchant availability (e.g. a kedai-kopi merchant only
 * enables cash + QRIS Dynamic + GoPay) is layered on top of this
 * inventory in a later slice — this enum is the *superset* that
 * the platform supports, not the *subset* a particular merchant
 * has turned on.
 */
enum class PaymentMethod(
    /** Stable code persisted to the backend (snake-screaming). */
    val code: String,
    /** Indonesian display label shown in the picker. */
    val displayLabel: String,
    /** `true` if the method needs a live HTTP round-trip to settle. */
    val requiresOnline: Boolean,
) {
    CASH(code = "CASH", displayLabel = "Tunai", requiresOnline = false),
    EDC(code = "EDC", displayLabel = "Kartu (EDC)", requiresOnline = false),
    QRIS_STATIC(code = "QRIS_STATIC", displayLabel = "QRIS Statis", requiresOnline = false),
    QRIS_DYNAMIC(code = "QRIS_DYNAMIC", displayLabel = "QRIS Dinamis", requiresOnline = true),
    GOPAY(code = "GOPAY", displayLabel = "GoPay", requiresOnline = true),
    OVO(code = "OVO", displayLabel = "OVO", requiresOnline = true),
    DANA(code = "DANA", displayLabel = "DANA", requiresOnline = true),
    SHOPEEPAY(code = "SHOPEEPAY", displayLabel = "ShopeePay", requiresOnline = true),
    LINKAJA(code = "LINKAJA", displayLabel = "LinkAja", requiresOnline = true),
    BANK_TRANSFER(code = "BANK_TRANSFER", displayLabel = "Transfer Bank", requiresOnline = false),
    CREDIT(code = "CREDIT", displayLabel = "Piutang", requiresOnline = false),
    DEPOSIT(code = "DEPOSIT", displayLabel = "Deposit", requiresOnline = false),
    VOUCHER(code = "VOUCHER", displayLabel = "Voucher", requiresOnline = false),
    LOYALTY_POINT(code = "LOYALTY_POINT", displayLabel = "Poin Loyalty", requiresOnline = false),
    OTHER(code = "OTHER", displayLabel = "Lainnya", requiresOnline = false),
    ;

    companion object {
        /**
         * Resolve a code string back to its enum entry.
         *
         * Returns `null` for an unknown code rather than throwing
         * so the caller can decide whether an unfamiliar code is a
         * hard error (e.g. a bug to surface) or a soft fallback
         * (e.g. show "Lainnya" with the raw label as a hint).
         */
        fun fromCode(code: String): PaymentMethod? =
            entries.firstOrNull { it.code == code }
    }
}
